plugins/SpectrumAnalyzer/SaControlsDialog.cpp

```diff
@@ -0,0 +1,227 @@
+
+	ComboBox *ampRangeCombo = new ComboBox(this, tr("Amplitude range"));
+	ampRangeCombo->setToolTip(tr("Amplitude range"));
+	ampRangeCombo->setMinimumSize(100, 22);
+	ampRangeCombo->setMaximumSize(200, 22);
+	ampRangeCombo->setModel(&controls->m_ampRangeModel);
+	config_layout->addWidget(ampRangeCombo, 2, 3, 2, 1);
+
+	// FFT: block size: icon and selector
+	QLabel *blockSizeLabel = new QLabel("", this);
+	QPixmap *blockSizeIcon = new QPixmap(PLUGIN_NAME::getIconPixmap("block_size"));
+	blockSizeIcon->setDevicePixelRatio(devicePixelRatio());
+	blockSizeLabel->setPixmap(blockSizeIcon->scaled(iconSize, iconSize, Qt::IgnoreAspectRatio, Qt::SmoothTransformation));
+	config_layout->addWidget(blockSizeLabel, 0, 4, 2, 1, Qt::AlignRight);
+
+	ComboBox *blockSizeCombo = new ComboBox(this, tr("FFT block bize"));
+	blockSizeCombo->setToolTip(tr("FFT block size"));
+	blockSizeCombo->setMinimumSize(100, 22);
+	blockSizeCombo->setMaximumSize(200, 22);
+	blockSizeCombo->setModel(&controls->m_blockSizeModel);
+	config_layout->addWidget(blockSizeCombo, 0, 5, 2, 1);
+	processor->reallocateBuffers();
+	connect(&controls->m_blockSizeModel, &ComboBoxModel::dataChanged, [=] {processor->reallocateBuffers();});
+
+	// FFT: window type: icon and selector
+	QLabel *windowLabel = new QLabel("", this);
+	QPixmap *windowIcon = new QPixmap(PLUGIN_NAME::getIconPixmap("window"));
+	windowIcon->setDevicePixelRatio(devicePixelRatio());
+	windowLabel->setPixmap(windowIcon->scaled(iconSize, iconSize, Qt::IgnoreAspectRatio, Qt::SmoothTransformation));
+	config_layout->addWidget(windowLabel, 2, 4, 2, 1, Qt::AlignRight);
+
+	ComboBox *windowCombo = new ComboBox(this, tr("FFT window type"));
+	windowCombo->setToolTip(tr("FFT window type"));
+	windowCombo->setMinimumSize(100, 22);
+	windowCombo->setMaximumSize(200, 22);
+	windowCombo->setModel(&controls->m_windowModel);
+	config_layout->addWidget(windowCombo, 2, 5, 2, 1);
+	processor->rebuildWindow();
+	connect(&controls->m_windowModel, &ComboBoxModel::dataChanged, [=] {processor->rebuildWindow();});
+
+
+	// QSplitter middle and bottom: spectrum display widgets
+	m_spectrum = new SaSpectrumView(controls, processor, this);
+	display_splitter->addWidget(m_spectrum);
+
+	m_waterfall = new SaWaterfallView(controls, processor, this);
+	display_splitter->addWidget(m_waterfall);
+	m_waterfall->setVisible(m_controls->m_waterfallModel.value());
+	connect(&controls->m_waterfallModel, &BoolModel::dataChanged, [=] {m_waterfall->updateVisibility();});
+}
+
+
+// Suggest the best current widget size.
+QSize SaControlsDialog::sizeHint() const
+{
+	// Best width is determined by spectrum display sizeHint.
+	// Best height depends on whether waterfall is visible and
+	// consists of heights of the config section, spectrum, waterfall
+	// and some reserve for margins.
+	if (m_waterfall->isVisible())
+	{
+		return QSize(m_spectrum->sizeHint().width(),
+					 m_configHeight + m_spectrum->sizeHint().height() + m_waterfall->sizeHint().height() + 50);
```

Comment:
This line is also too long, and it can be put together with the `else` case (`return QSize( ... , ... + (m_waterfall->isVisible() ? ... : ...)`). The ternary operator is matter of taste, though.

Reply:
Essentially discussed one comment / review above? I would prefer keeping it as it is, at it is clear at a glance what's the difference between the two cases.

I like ternary operators, but at this point it would feel more like cheating to avoid the extra lines caused by the opening-bracket-at-new line convention. :D And it would probably make the line even longer and less readable..

Reply:
OK, I think we can resolve this.